src/neuralnet/tally.cpp

```diff
@@ -0,0 +1,1190 @@
+#include "beacon.h"
+#include "main.h"
+#include "neuralnet/cpid.h"
+#include "neuralnet/researcher.h"
+#include "neuralnet/tally.h"
+#include "util.h"
+
+#include <deque>
+#include <unordered_map>
+
+using namespace NN;
+
+namespace {
+//!
+//! \brief Number of days that the tally scans backward from to calculate
+//! the average network payment.
+//!
+constexpr size_t TALLY_DAYS = 14;
+
+//!
+//! \brief Used to convert elapsed accrual seconds into days, so we declare it
+//! as a double type.
+//!
+constexpr double ONE_DAY_IN_SECONDS = 86400.0;
+
+//!
+//! \brief Round a value to intervals of 0.025.
+//!
+//! Used to calculate the network magnitude unit.
+//!
+double SnapToGrid(double d)
+{
+    double dDither = .04;
+    double dOut = RoundFromString(RoundToString(d * dDither, 3), 3) / dDither;
+    return dOut;
+}
+
+//!
+//! \brief Get the multiplier used to calculate the maximum research reward
+//! amount.
+//!
+//! Survey Results: Start inflation rate: 9%, end=1%, 30 day steps, 9 steps,
+//! mag multiplier start: 2, mag end .3, 9 steps
+//!
+//! \return A value in units of GRC (not COIN) that represents the maximum
+//! research reward expected per block.
+//!
+int64_t GetMaxResearchSubsidy(const int64_t nTime)
+{
+    // Gridcoin Global Daily Maximum Researcher Subsidy Schedule
+    int MaxSubsidy = 500;
+
+    if (nTime > 1447977700)                              MaxSubsidy =  50; // from  11-20-2015 forever
+    else if (nTime >= 1445309276 && nTime <= 1447977700) MaxSubsidy =  75; // between 10-20-2015 and 11-20-2015
+    else if (nTime >= 1438310876 && nTime <= 1445309276) MaxSubsidy = 100; // between 08-01-2015 and 10-20-2015
+    else if (nTime >= 1430352000 && nTime <= 1438310876) MaxSubsidy = 150; // between 05-01-2015 and 07-31-2015
+    else if (nTime >= 1427673600 && nTime <= 1430352000) MaxSubsidy = 200; // between 03-30-2015 and 04-30-2015
+    else if (nTime >= 1425254400 && nTime <= 1427673600) MaxSubsidy = 250; // between 02-28-2015 and 03-30-2015
+    else if (nTime >= 1422576000 && nTime <= 1425254400) MaxSubsidy = 300; // between 01-30-2015 and 02-28-2015
+    else if (nTime >= 1419897600 && nTime <= 1422576000) MaxSubsidy = 400; // between 12-30-2014 and 01-30-2015
+    else if (nTime >= 1417305600 && nTime <= 1419897600) MaxSubsidy = 400; // between 11-30-2014 and 12-30-2014
+    else if (nTime >= 1410393600 && nTime <= 1417305600) MaxSubsidy = 500; // between inception  and 11-30-2014
+
+    // The .5 allows for fractional amounts after the 4th decimal place (used to store the POR indicator)
+    return MaxSubsidy+.5;
+}
+
+//!
+//! \brief Organizes superblocks for lookup to calculate research rewards.
+//!
+class SuperblockIndex
+{
+    //!
+    //! \brief
+    //!
+    static constexpr size_t CACHE_SIZE = 3;
```

Comment:
Yep...this class will need much more intelligence to enable superblock window accrual calculations to perform optimally. We will need to revisit this in the PR that implements those. For now, the cache is designed to avoid loading superblocks from disk at each tally averaging interval like the old code does and allows the tally to avoid rescanning the chain for superblocks around 90% of the time.